### DISCUSSION TITLE
Messages are getting dropped because message queue is full

### DISCUSSION THREAD

**Discussion by karanbangia1998** [emqx/emqx#9335] [Q&A] [answered]
We recently saw this metric count of dropped.queue_full is increasing
-  publisher is publishing with QOS 2 and client is receiving with QOS 1
- we are not sending retained flag as true in any of the messages
can we know the potential reasons for this increasing number
![Screenshot 2022-11-02 at 4 40 54 PM (1)](https://user-images.githubusercontent.com/95840238/200875114-436aa3c6-8c36-4e3a-afc9-3868182f7215.png)

**Comment by lafirest** [comment, marked as answer]
@karanbangia1998 
Hi, when the incoming rate is bigger than the consuming rate, the message which can't consume ASAP will be stored in the mqueue and when this queue is full, the oldest message will be dropped.
There are some options for you:
1. downgrade the publish QoS to 1
2. improve the consuming rate, e.g. use the [shared subscription](https://www.emqx.io/docs/en/v4.3/advanced/shared-subscriptions.html)
3. increase the [maximum length of the mquem](https://www.emqx.io/docs/en/v4.3/advanced/inflight-window-and-message-queue.html#inflight-queue-and-receive-maximum)

**Reply by karanbangia1998** [reply]
if we are sending messages to the topic where client is not active and that is large in number, can this be cause of issue?
and is there any way to debug it..like a particular topic for which it is happening?

**Reply by lafirest** [reply]
Yes, it will be a problem if you don't consume quickly the message whose QoS is 1 or 2. 
The subscriber should respond ASAP or these messages may be dropped. 
The [System Topics](https://www.emqx.io/docs/en/v4.3/advanced/system-topic.html#cluster-status-information), [Metrics](https://www.emqx.io/docs/en/v4.3/advanced/metrics-and-stats.html#metrics-stats), [Log Trace](https://www.emqx.io/docs/en/v4.3/getting-started/log.html), and [Slow Subscriptions](https://www.emqx.com/en/blog/emqx-v-4-4-new-feature-a-top-view-of-slowest-subscribers) may can help you.

**Reply by karanbangia1998** [reply]
lets suppose client A is not connected to MQTT and we are sending thousands of messages to client 
A, can this scenario affect client B's messages getting dropped because queue is full?
thanks for replying @lafirest

**Reply by lafirest** [reply]
No, it will not.
Each connection/session has an isolated mqueue &&  inflight queue, they don't interfere with each other.
The only reason for clients' messages getting dropped is the slow acknowledgment response. 
The `slow` is relative to the message incoming rate not saying it is really slow.
